pkg/scheduler/util/utils.go

```diff
@@ -140,3 +140,32 @@ func MoreImportantPod(pod1, pod2 interface{}) bool {
 	}
 	return GetPodStartTime(pod1.(*v1.Pod)).Before(GetPodStartTime(pod2.(*v1.Pod)))
 }
+
+// NonBlockingSendError sends val to the provided channel without blocking on the receiver.
+// Note that this function should only be used for errors since type casting the interface
+// adds overhead that is best avoided in performance critical paths.
+func NonBlockingSendError(ch chan<- interface{}, val interface{}) {
+	select {
+	case ch <- val:
+	default:
+	}
+}
+
+// NonBlockingSendErrorWithCancel similar to NonBlockingSendError, but allows passing
+// a cancel function that will be called after sending the error.
+func NonBlockingSendErrorWithCancel(ch chan<- interface{}, val interface{}, cancel context.CancelFunc) {
```

Comment:
I don't think it is a good idea to have this function, `NonBlockingSendError` is a well-defined function with a specific purpose. Adding an extra cancel argument does not sound good.

Reply:
Can you explain why adding another argument "does not sound good"? The function semantics are clearly defined both in the function name and the comment: 1) send an error, 2) cancel the context, two things that happen hand in hand in many places. 

I actually think this is good practice, it will allow us to enforce a policy to call cancel to exit parallel loops early in case of errors.

Reply:
> Can you explain why adding another argument "does not sound good"? The function semantics are clearly defined both in the function name and the comment: 1) send an error, 2) cancel the context, two things that happen hand in hand in many places.

I'm not saying the function name isn't self-explanatory. 

> I actually think this is good practice, it will allow us to enforce a policy to call cancel to exit parallel loops early in case of errors.

Send errors and cancel context (cleanup) is two different functions. There are scenarios, we don't have to call cancel.

Add `NonBlockingSendErrorWithCancel` with an extra argument which has weak cohesion with the original purpose (send errors) does not sound right to me in this case. It would work in some cases, but `NonBlockingSendError` could meet most of our requirements. I don't feel strongly about this, but we need functions with a proper level of abstraction in utils.

Reply:
> > Can you explain why adding another argument "does not sound good"? The function semantics are clearly defined both in the function name and the comment: 1) send an error, 2) cancel the context, two things that happen hand in hand in many places.
> 
> I'm not saying the function name isn't self-explanatory.
> 
> > I actually think this is good practice, it will allow us to enforce a policy to call cancel to exit parallel loops early in case of errors.
> 
> Send errors and cancel context (cleanup) is two different functions. There are scenarios, we don't have to call cancel.

That is why we have NonBlockingSendError.

> 
> Add `NonBlockingSendErrorWithCancel` with an extra argument which has weak cohesion with the original purpose (send errors) does not sound right to me in this case. It would work in some cases, but `NonBlockingSendError` could meet most of our requirements, and it's better to do the cleanup outside of `NonBlockingSendError`.

Recording an error and executing a callback (cancel in this case) is a very common pattern, I don't think this "has weak cohesion".

> 
> I don't feel strongly about this,

Sounds good, because I believe it is useful. 

> but we need functions with a proper level of abstraction in utils.

I don't mind encapsulating all of that in a new type named something like ErrorRecorder as you suggested in the other PR, that is even better.

Reply:
> Recording an error and executing a callback (cancel in this case) is a very common pattern, I don't think this "has weak cohesion".

Executing a callback could be a common pattern but I don't think to add a `context.CancelFunc` is common. Even if it is common, we could also do the cleanup after sending errors.

Reply:
> > Recording an error and executing a callback (cancel in this case) is a very common pattern, I don't think this "has weak cohesion".
> 
> Executing a callback could be a common pattern but I don't think to add a `context.CancelFunc` is common. 

cancel is a callback!

> Even if it is common, we could also do the cleanup after sending errors.

Putting it here avoids repeating the call across the code.

Reply:
I found similar logic here in these places which do not need to call `CancelFunc`, and I still think it should not be part of the API.

+ https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/priorities/interpod_affinity.go#L194-L218 
+ https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L760-L782
+ https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1563-L1596

And we probably go too far on this, use select and channel to capture the first error is quite common in the codebase, I use following cmd and find many similar logics.